conn.go

```diff
@@ -694,7 +718,7 @@ func (c *Conn) recv(ctx context.Context) error {
 	} else if head.stream == -1 {
 		// TODO: handle cassandra event frames, we shouldnt get any currently
 		framer := newFramer(c.compressor, c.version)
-		if err := framer.readFrame(c, &head); err != nil {
+		if err := framer.readFrame(r, &head); err != nil {
 			return err
 		}
 		go c.session.handleEvent(framer)
```

Comment:
> This should be changed to r as well right?

You're right. I didn't notice it here.

> I think this is even the case on readUncompressedFrame and readCompressedFrame... It's using c.r directly but it should be using c so the additional logic from Conn.Read() is used right?

I didn't even notice that `Conn` implements `io.Reader`. I assumed that `*bufio.Reader` should be used to read from the socket because I had seen [line 673](https://github.com/apache/cassandra-gocql-driver/blob/953e0df999cabb3f5eef714df9921c00e9f632c2/conn.go#L673C1-L674C1) where `c.r` is used:
```
head, err := readHeader(c.r, c.headerBuf[:])
```

> I think we should create a type (maybe connReader?) that wraps the *bufio.Reader object and then move the implementation of Conn.Read() to that new type so we ensure that we only have one object that implements io.Reader. This way c.r becomes an object of *connReader instead of *bufio.Reader, Conn no longer implements io.Reader and everytime we want to actually read from the socket we have to use the c.r field instead of c directly.

Something like [this](https://github.com/worryg0d/gocql/commit/44cc2fc9f2344928f261d43c60071db1170d706c)? It makes sense since it forces us to use `c.r` instead of `Conn` itself. However, there is `net.Conn` still presented in `Conn`, which may be accidentally used instead.

Reply:
> Something like [this](https://github.com/worryg0d/gocql/commit/44cc2fc9f2344928f261d43c60071db1170d706c)? It makes sense since it forces us to use c.r instead of Conn itself. However, there is net.Conn still presented in Conn, which may be accidentally used instead.

Yeah, we can go even further and completely move both `net.Conn` and `bufio.Reader` fields to a separate type that implements the `net.Conn` interface. the `timeout` field could also be removed from `gocql.Conn`. This way we ensure nobody uses the underlying `net.Conn` object directly by mistake.

```
type ConnWrapper interface {
    net.Conn
}

type connWrapper struct {
	conn      net.Conn
	r         *bufio.Reader
	timeout   time.Duration
}

func (c *connWrapper) Read(p []byte) (n int, err error) {
	const maxAttempts = 5
	for i := 0; i < maxAttempts; i++ {
		var nn int
		if c.timeout > 0 {
			c.SetReadDeadline(time.Now().Add(c.timeout))
		}
		nn, err = io.ReadFull(c.r, p[n:])
		n += nn
		if err == nil {
			break
		}
		if verr, ok := err.(net.Error); !ok || !verr.Temporary() {
			break
		}
	}
	return
}

// implement remaining net.Conn`functions by just delegating to c.conn
```

Reply:
I updated the impl of [ConnReader](https://github.com/worryg0d/gocql/commit/bc29d6e491e75fd72f0c77445151dc117d62ac02). The only thing I don't like here is that [calling](https://github.com/worryg0d/gocql/blob/bc29d6e491e75fd72f0c77445151dc117d62ac02/conn.go#L567C1-L569C2)`Close()` on the `r` actually closes the whole `net.Conn`. We should probably move `contextWriter` from `Conn` to the wrapper as well...

Reply:
> Hmm this is the existing behavior though right?

Right, it calls close of underlying conn, so the behavior is not changed. It's just a bit confusing for me that calling `Close()` on `ConnReader` affects `writerContext` as well because they both use the same `net.Conn`.

> One thing we can do is remove the lower case Conn.close() because at the moment we have a Conn.Close() and a Conn.close(). The lower case one is just confusing, the Conn.Close() function should just call connReader.Close() 
directly.

I removed `Conn.close()` and added conn reader to this PR [a79c985](https://github.com/worryg0d/gocql/commit/a79c9853a50f2c6e2afc262be3a241e1bc1a97ff)